jenkinsci/pipeline-maven-plugin
# WIP [JENKINS-44721] Trigger pipelines that depend on generated snapshot artifacts

[JENKINS-44721](https://issues.jenkins-ci.org/browse/JENKINS-44721) Trigger pipelines that depend on generated snapshot artifacts

# TODO

* [ ] better handling of database error: add a circuit breaker
* [x] verify the security logic in [org.jenkinsci.plugins.pipeline.maven.listeners.DownstreamPipelineTriggerRunListener#onCompleted](https://github.com/jenkinsci/pipeline-maven-plugin/blob/0291bb0f74e7c1332a33c7e8990363edb18de056/jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/listeners/DownstreamPipelineTriggerRunListener.java). There are cases, at least with CloudBees RBAC authorization, were the downstream job is not triggered for security reasons that seem to be invalid (see [here](https://github.com/jenkinsci/pipeline-maven-plugin/blob/0291bb0f74e7c1332a33c7e8990363edb18de056/jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/listeners/DownstreamPipelineTriggerRunListener.java#L97))
* [ ] support an external H2 database instead of the inprocess h2 (may be done in a subsequent version)
* [ ] support PostgreSQL (will be done in a subsequent version)

# Testing the downstream trigger

## Testing dataset

* Upstream job: simple pipeline job `my-jar` using the script 
```
node {
    stage ("Build") {
        git "https://github.com/cyrille-leclerc/my-jar.git"
        withMaven(maven:"maven-3.5.0") {
            sh "mvn clean install"
        }
    }
}
```

* Downstream job: multibranch pipeline job `my-war` for https://github.com/cyrille-leclerc/my-war

## Testing steps

* Launch Jenkins, maybe `mvn hpi:run` from your IDE to use a debugger
* Enable logs on the pipeline-maven-plugin
* Launch `my-jar`
   * The generated jar gets registered in the database of maven artifacts
* Launch `my-war`
   * The jar `my-jar` gets registered as a dependency of `my-war.war`
* Launch `my-jar`, the build should trigger a downstream build of `my-war/master` and the build logs of `my-jar` should display, at the end of the logs 
```
[withMaven] Scheduling downstream pipeline my-war » master...
```
